ios/PacketTunnel/PacketTunnelProvider.swift

```diff
@@ -307,6 +322,144 @@ class PacketTunnelProvider: NEPacketTunnelProvider, TunnelMonitorDelegate {
+        case unableToReadCredentials
+        case fetchingAccountDataFinishedWith(error: Error)
+        case fetchingDeviceDataFinishedWith(error: REST.Error)
+        case ranOutOfTime(newExpiry: Date)
+    }
+
+    private func createGetAccountDataOperation(accountNumber: String)
+        -> ResultOperation<REST.AccountData, REST.Error>
+    {
+        let operation = ResultBlockOperation<REST.AccountData, REST.Error>(
+            dispatchQueue: DispatchQueue(label: "PacketTunnel.AccountDataProxy", qos: .utility)
+        )
+
+        operation.setExecutionBlock { operation in
+            let task = self.accountsProxy.getAccountData(
+                accountNumber: accountNumber,
+                retryStrategy: .default
+            ) { completion in
+                operation.finish(completion: completion)
+            }
+
+            operation.addCancellationBlock {
+                task.cancel()
+            }
+        }
+
+        return operation
+    }
+
+    private func createGetDeviceDataOperation(accountNumber: String, identifier: String)
+        -> ResultOperation<REST.Device, REST.Error>
+    {
+        let dispatchQueue = DispatchQueue(label: "PacketTunnel.DeviceDataProxy", qos: .utility)
+        let operation = ResultBlockOperation<REST.Device, REST.Error>(dispatchQueue: dispatchQueue)
+
+        operation.setExecutionBlock { operation in
+            let task = self.deviceProxy.getDevice(
+                accountNumber: accountNumber,
+                identifier: identifier,
+                retryStrategy: .default
+            ) { completion in
+                operation.finish(completion: completion)
+            }
+
+            operation.addCancellationBlock {
+                task.cancel()
+            }
+        }
+
+        return operation
+    }
+
+    private func verifyData(
+        accountNumber: String,
+        deviceIdentifier: String,
+        completion: @escaping (FailedToEstablishConnectionError?) -> Void
+    ) {
+        let accountOperation = createGetAccountDataOperation(accountNumber: accountNumber)
+        let deviceOperation = createGetDeviceDataOperation(
+            accountNumber: accountNumber,
+            identifier: deviceIdentifier
+        )
+
+        let completionOperation = AsyncBlockOperation(dispatchQueue: .global()) {
+            let accountCompletion = accountOperation.completion
+            let deviceCompletion = deviceOperation.completion
+
+            if let accountData = accountCompletion?.value {
+                if accountData.expiry > Date() {
+                    completion(.ranOutOfTime(newExpiry: accountData.expiry))
+                }
+            }
+
+            if let deviceCompletionError = deviceCompletion?.error {
+                completion(.fetchingDeviceDataFinishedWith(error: deviceCompletionError))
+            }
+        }
+
+        completionOperation.addDependencies([accountOperation, deviceOperation])
```

Comment:
Forgot to add operations to `AsyncOperationQueue`?

Reply:
Done.